lib/GPU/IR/GPUDialect.cpp

```diff
@@ -220,12 +220,13 @@ parseSizeAssignment(OpAsmParser *parser,
                     MutableArrayRef<OpAsmParser::OperandType> sizes,
                     MutableArrayRef<OpAsmParser::OperandType> regionSizes,
                     MutableArrayRef<OpAsmParser::OperandType> indices) {
-  if (parser->parseLParen() || parser->parseRegionArgument(indices[0]) ||
-      parser->parseComma() || parser->parseRegionArgument(indices[1]) ||
-      parser->parseComma() || parser->parseRegionArgument(indices[2]) ||
-      parser->parseRParen() || parser->parseKeyword("in") ||
-      parser->parseLParen())
+  assert(indices.size() == 3 && "space for three indices expected");
+  SmallVector<OpAsmParser::OperandType, 3> args;
+  if (parser->parseRegionArgumentList(args, /*requiredOperandCount=*/3,
+                                      OpAsmParser::Delimiter::Paren) ||
+      parser->parseKeyword("in") || parser->parseLParen())
     return failure();
+  std::copy(args.begin(), args.end(), indices.begin());
```

Comment:
Could this copy be avoided by parsing into indices directly?

Reply:
Unfortunately, no AFAIU. The 'indices' is a MutableArrayRef, while the parse method takes a SmallVector. The former's length can't be changed although its contents can be, and so there can't be an implicit conversion to the latter.

Reply:
Can you make this function take `SmallVectorImpl &` instead, forward it to the parse method, and then concatenate vectors using move iterators at the top call site?

Reply:
The declaration and concatenation of six different vectors is going to be just too verbose at the call site! Did you want that?

Reply:
Also, IIUC, using move iterators to concatenate three SmallVectors won't prevent any copies - in fact, it'll lead to 14 copies here instead of the six right now. (SmallVectors store elements contiguously in a single allocation.) Using move on the OperandType itself isn't more efficient since all its fields are POD or POD-like, and one could use move anyway separately on these three here.